pkg/kube/krt/README.md

```diff
@@ -0,0 +1,88 @@
+# `krt`: **K**ubernetes Declarative Controller **R**un**t**ime
+
+`krt` provides a framework for building _declarative_ controllers.
+See the [design doc](https://docs.google.com/document/d/1-ywpCnOfubqg7WAXSPf4YgbaFDBEU9HIqMWcxZLhzwE/edit#heading=h.ffjmk8byb9gt) and [KubeCon talk](https://sched.co/1R2oY) for more background.
+
+The framework aims to solve a few problems with writing controllers:
+* Operate on any types, from any source. Kubernetes provides informers, but these only work on Kubernetes types read from Kubernetes objects.
+  * `krt` can accept any object type, from any source, and handle them the same.
+* Provide high level abstractions.
+  * Controller authors can write simple transformation functions from `Input` -> `Output` (with dependencies); the framework handles all the state automatically.
+
+## Key Primitives
+
+The most important primitive provided is the `Collection` interface.
+This is basically an `Informer`, but not tied to Kubernetes.
+
+Currently, there are three ways to build a `Collection`:
+* Built from an `Informer` with `WrapClient` or `NewInformer`.
+* Statically configured with `NewStatic`.
+* Derived from other collections (more information on this below).
+
+Unlike `Informers`, these primitives work on arbitrary objects.
+However, these objects are expected to have some properties, depending on their usage.
+These are *not* expressed as generic constraints due to limitations in Go's type system.
+
+* Each object `T` must have a unique `Key[T]` (which is just a typed wrapper around `string`) that uniquely identifies the object.
+    Default implementations exist for Kubernetes objects, Istio `config.Config` objects, and `ResourceName() string` implementations.
+* `Equals(k K) bool` may be implemented to provide custom implementations to compare objects. Comparison is done to detect if changes were made.
+  Default implementations are available for Kubernetes and protobuf objects, and will fallback to `reflect.DeepEqual`.
+* A `Name`, `Namespace`, `Labels`, and `LabelSelector` may optionally be included, for use with filters (see below).
+
+## Derived Collections
+
+The core of the framework is in the ability to derive collections from others.
+
+In general, these are built by providing some `func(inputs...) outputs...`.
+While more could be expressed, there are currently three forms implemented.
+
+* `func() *O` via `NewSingleton`
+  * This generates a collection that has a single value. An example would be some global configuration.
+* `func(input I) *O` via `NewCollection`
+  * This generates a one-to-one mapping of input to output. An example would be a transformation from a `Pod` type to a generic `Workload` type.
+* `func(input I) []O` via `NewManyCollection`
```

Comment:
Are there any constraints or requirements on the "many" return type? Must it be sorted, a set, etc? Does it make sense to allow explicit declaration of these constraints?